Rewrite Onboard without unstated or Router

GAumala opened the pull request:
Keep session state (is logged in?) in App component.
Create new stateful component Onboard which keeps track of the onboard
steps and input data and renders the correct form at each step.

I removed unstated and react-router because they were unnecessarily complex. You do't need routes for sign up, If you add them you will have to sprinkle redirects all over your code. I don't like unstated because updating multiple stores from any component can become a mess very quickly.

I still need to add validations on the forms and http requests.

mysticmode commented:
So, with the current code you've done, How can we update the browser URL?

I can understand that react-router-dom and redirects can get messy. But would you recommend to change the URL using pushState?

The steps has to be in it's each URL as per the current design I've made.

mysticmode commented on src/onboard/Onboard.js:
```diff
@@ -0,0 +1,332 @@
+
+
+  skipSMTPConfigStep = () => 
+    this.setState({ smtp: null, step: Step.storage });
```
What does `step: Step.storage` do? Should it be `Step.setupStorage`?

Also, when I click 'Skip' button it says this in the console
```
Form submission canceled because the form is not connected
```

mysticmode commented:
All looks good! I just need to about those changes.

GAumala commented on src/App.js:
```diff
@@ -1,51 +1,42 @@
       <div className="app">
         <div id="loader"><i></i></div>
-        <Header uploadBooksAPI={apiRoutes.uploadBooks} />
+        <Header uploadBooksAPI="/upload-books" />
```
I'd rather keep those routes in `client` object that encapsulates all the http logic. These thinks don't change at runtime so they should not be passed as props.

GAumala commented on src/onboard/Onboard.js:
```diff
@@ -0,0 +1,332 @@
+    <input type="submit" className="button button--primary onboard__submit" value="Submit" />
+  </form>
+
+export default class OnboardContainer extends Component {
```
I'm thinking about moving the stateless components to different files.

mysticmode commented on src/onboard/validation.js:
```diff
@@ -0,0 +1,102 @@
+export const validateSMTPConfig = smtp => {
+  const hostname = validateHostname(smtp.hostname)
+  const username = validateSMTPUsername(smtp.username)
+  const { value: newPort, error: portError } = validatePort(smtp.port)
```
I don't understand this, sorry! So `validatePort` returns an object and what does `const { value: newPort, error: portError }` does?

mysticmode commented on src/onboard/validation.js:
```diff
@@ -0,0 +1,102 @@
+export const validateSMTPConfig = smtp => {
+  const hostname = validateHostname(smtp.hostname)
+  const username = validateSMTPUsername(smtp.username)
+  const { value: newPort, error: portError } = validatePort(smtp.port)
```
I don't understand this, sorry! `validatePort` returns an object and what does this `const { value: newPort, error: portError }` do?

mysticmode replied:
Ah I see, those are constants, never seen that approach before. is it from ES6? I couldn't search for this and find articles.

GAumala replied:
Yeah it's [ES6 destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names).

mysticmode commented on src/onboard/Onboard.js:
```diff
@@ -0,0 +1,352 @@
+  handleSMTPInputChange = ({ target }) => {
+    const { name, value } = target;
+    this.setState(prevState => {
+      const prevSmtp = prevState.smtp || defaultState.smtp;
```
What does `prevState.smtp || defaultState.smtp` do?

GAumala replied:
if `prevState.smtp` is falsy (`undefined`, `null`, `false`, `0`, `""`), then `defaultState.smtp` is assigned to `prevSmtp` instead of `prevState.smtp`.

I think I don't need this check anymore, so I'll probably remove it.

GAumala commented:
Added some tests. It's not necessary to test absolutely everything, only the general things and stuff that worries you the most. Having a good test suite is good because you can try stuff with mocked API calls, which means that you can develop without having to start the actual joyread server. This should make it easier for new contributors that only do frontend.

some validation functions might grow bigger so we could add individual tests for those.

The key pattern is to separate stateful containers from the stateless ones. We test the stateful components by mocking the stateless components and inspecting how the passed props change as we interact with it.

I believe that files should be grouped by feature instead of their type. All files needed by `OnboardContainer` are right next to it inside the `onboard` directory. If the storage form got bigger, for example we added more forms for integration with more services like MEGA or Dropbox, then those could go in a new subdirectory `onboard/storage`. It would be great. Any css used exclusively for onboard forms should go here too.
